File: fsharp/space-age/SpaceAge.fs
module SpaceAge

type Planet =
    | Earth
    | Mercury
    | Venus
    | Mars
    | Jupiter
    | Saturn
    | Uranus
    | Neptune

let age (planet: Planet) (seconds: int64): float =
    let secondsInYear = 60.0 * 60.0 * 24.0 * 365.25
    let earthYears = float seconds / secondsInYear

    match planet with
    | Earth     -> earthYears
    | Mercury   -> earthYears / 0.2408467
    | Venus     -> earthYears / 0.61519726
    | Mars      -> earthYears / 1.8808158
    | Jupiter   -> earthYears / 11.862615
    | Saturn    -> earthYears / 29.447498
    | Uranus    -> earthYears / 84.016846
    | Neptune   -> earthYears / 164.79132

File: clojure/space-age/src/space_age.clj
(ns space-age)

(def earth-orbit-in-seconds 31557600)

(def orbital-periods
  {"mercury" 0.2408467
   "venus"   0.61519726
   "earth"   1
   "mars"    1.8808158
   "jupiter" 11.862615
   "saturn"  29.447498
   "uranus"  84.016846
   "neptune" 164.79132})

(defmacro create-orbital-fns []
  (let [fns (for [[planet period] orbital-periods]
              `(defn ~(symbol (str "on-" planet))
                 [seconds#]
                 (/ seconds# earth-orbit-in-seconds ~period)))]
    `(do ~@fns)))

(create-orbital-fns)
